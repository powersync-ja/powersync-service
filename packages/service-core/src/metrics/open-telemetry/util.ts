import { OTLPMetricExporter } from '@opentelemetry/exporter-metrics-otlp-http';
import { PrometheusExporter } from '@opentelemetry/exporter-prometheus';
import { MeterProvider, MetricReader, PeriodicExportingMetricReader } from '@opentelemetry/sdk-metrics';
import { logger } from '@powersync/lib-services-framework';
import { ServiceContext } from '../../system/ServiceContext.js';
import { MetricsFactory } from '../metrics-interfaces.js';
import { OpenTelemetryMetricsFactory } from './OpenTelemetryMetricsFactory.js';

import pkg from '../../../package.json' with { type: 'json' };
import { resourceFromAttributes } from '@opentelemetry/resources';

export function createOpenTelemetryMetricsFactory(context: ServiceContext): MetricsFactory {
  const { configuration, lifeCycleEngine, storageEngine } = context;
  const configuredExporters: MetricReader[] = [];

  if (configuration.telemetry.prometheus_port) {
    const prometheusExporter = new PrometheusExporter({
      port: configuration.telemetry.prometheus_port,
      preventServerStart: true
    });
    configuredExporters.push(prometheusExporter);

    lifeCycleEngine.withLifecycle(prometheusExporter, {
      start: async () => {
        await prometheusExporter.startServer();
        logger.info(`Prometheus metric export enabled on port:${configuration.telemetry.prometheus_port}`);
      }
    });
  }

  if (!configuration.telemetry.disable_telemetry_sharing) {
    const periodicExporter = new PeriodicExportingMetricReader({
      exporter: new OTLPMetricExporter({
        url: configuration.telemetry.internal_service_endpoint
      }),
      exportIntervalMillis: 1000 * 60 * 5 // 5 minutes
    });

    configuredExporters.push(periodicExporter);
  }

  let resolvedInstanceId: (id: string) => void;
  const instanceIdPromise = new Promise<string>((resolve) => {
    resolvedInstanceId = resolve;
  });

  lifeCycleEngine.withLifecycle(null, {
    start: async () => {
      const bucketStorage = storageEngine.activeBucketStorage;
      try {
        const instanceId = await bucketStorage.getPowerSyncInstanceId();
        resolvedInstanceId(instanceId);
      } catch (err) {
        resolvedInstanceId('Unknown');
      }
    }
  });

  const resource = resourceFromAttributes({
    ['service']: 'PowerSync',
    ['service.version']: pkg.version,
    ['instance_id']: instanceIdPromise
  });

  // This triggers OpenTelemetry to resolve the async attributes (instanceIdPromise).
  // This will never reject, and we don't specifically need to wait for it.
  resource.waitForAsyncAttributes?.();

  const meterProvider = new MeterProvider({
    resource,

    readers: configuredExporters
  });

  lifeCycleEngine.withLifecycle(meterProvider, {
    stop: async () => {
      await meterProvider.shutdown();
    }
  });

  const meter = meterProvider.getMeter('powersync');

  return new OpenTelemetryMetricsFactory(meter);
}
