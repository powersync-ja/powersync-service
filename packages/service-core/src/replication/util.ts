import * as pgwire from '@powersync/service-jpgwire';

import * as util from '../util/util-index.js';
import { ReplicationColumn, ReplicationIdentity } from './PgRelation.js';

export interface ReplicaIdentityResult {
  columns: ReplicationColumn[];
  replicationIdentity: ReplicationIdentity;
}

export async function getPrimaryKeyColumns(
  db: pgwire.PgClient,
  relationId: number,
  mode: 'primary' | 'replident'
): Promise<ReplicationColumn[]> {
  const indexFlag = mode == 'primary' ? `i.indisprimary` : `i.indisreplident`;
  const attrRows = await util.retriedQuery(db, {
    statement: `SELECT a.attname as name, a.atttypid as typeid, a.attnum as attnum
                                    FROM pg_index i
                                             JOIN pg_attribute a
                                                  ON a.attrelid = i.indrelid
                                                      AND a.attnum = ANY (i.indkey)
                                    WHERE i.indrelid = $1::oid
                                      AND ${indexFlag}
                                      AND a.attnum > 0
                                      ORDER BY a.attnum`,
    params: [{ value: relationId, type: 'int4' }]
  });

  return attrRows.rows.map((row) => {
    return { name: row[0] as string, typeOid: row[1] as number };
  });
}

export async function getAllColumns(db: pgwire.PgClient, relationId: number): Promise<ReplicationColumn[]> {
  const attrRows = await util.retriedQuery(db, {
    statement: `SELECT a.attname as name, a.atttypid as typeid, a.attnum as attnum
                                    FROM pg_attribute a
                                    WHERE a.attrelid = $1::oid
                                      AND attnum > 0
                                    ORDER BY a.attnum`,
    params: [{ type: 'varchar', value: relationId }]
  });
  return attrRows.rows.map((row) => {
    return { name: row[0] as string, typeOid: row[1] as number };
  });
}

export async function getReplicationIdentityColumns(
  db: pgwire.PgClient,
  relationId: number
): Promise<ReplicaIdentityResult> {
  const rows = await util.retriedQuery(db, {
    statement: `SELECT CASE relreplident
        WHEN 'd' THEN 'default'
        WHEN 'n' THEN 'nothing'
        WHEN 'f' THEN 'full'
        WHEN 'i' THEN 'index'
     END AS replica_identity
FROM pg_class
WHERE oid = $1::oid LIMIT 1`,
    params: [{ type: 'int8', value: relationId }]
  });
  const idType: string = rows.rows[0]?.[0];
  if (idType == 'nothing' || idType == null) {
    return { replicationIdentity: 'nothing', columns: [] };
  } else if (idType == 'full') {
    return { replicationIdentity: 'full', columns: await getAllColumns(db, relationId) };
  } else if (idType == 'default') {
    return { replicationIdentity: 'default', columns: await getPrimaryKeyColumns(db, relationId, 'primary') };
  } else if (idType == 'index') {
    return { replicationIdentity: 'index', columns: await getPrimaryKeyColumns(db, relationId, 'replident') };
  } else {
    return { replicationIdentity: 'nothing', columns: [] };
  }
}
