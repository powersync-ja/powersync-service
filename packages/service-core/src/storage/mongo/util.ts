import { SqliteJsonValue } from '@powersync/service-sync-rules';
import * as bson from 'bson';
import * as crypto from 'crypto';
import * as mongo from 'mongodb';
import * as uuid from 'uuid';
import { OplogEntry } from '../../util/protocol-types.js';
import { ID_NAMESPACE, timestampToOpId } from '../../util/utils.js';
import { BucketDataDocument, ReplicaId } from './models.js';

/**
 * Lookup serialization must be number-agnostic. I.e. normalize numbers, instead of preserving numbers.
 * @param lookup
 */

export function serializeLookup(lookup: SqliteJsonValue[]) {
  const normalized = lookup.map((value) => {
    if (typeof value == 'number' && Number.isInteger(value)) {
      return BigInt(value);
    } else {
      return value;
    }
  });
  return new bson.Binary(bson.serialize({ l: normalized }));
}

export function idPrefixFilter<T>(prefix: Partial<T>, rest: (keyof T)[]): mongo.Condition<T> {
  let filter = {
    $gte: {
      ...prefix
    } as any,
    $lt: {
      ...prefix
    } as any
  };

  for (let key of rest) {
    filter.$gte[key] = new bson.MinKey();
    filter.$lt[key] = new bson.MaxKey();
  }

  return filter;
}

export function generateSlotName(prefix: string, sync_rules_id: number) {
  const slot_suffix = crypto.randomBytes(2).toString('hex');
  return `${prefix}${sync_rules_id}_${slot_suffix}`;
}

/**
 * Read a single batch of data from a cursor, then close it.
 *
 * We do our best to avoid MongoDB fetching any more data than this single batch.
 *
 * This is similar to using `singleBatch: true` in find options.
 * However, that makes `has_more` detection very difficult, since the cursor is always closed
 * after the first batch. Instead, we do a workaround to only fetch a single batch below.
 *
 * For this to be effective, set batchSize = limit in the find command.
 */
export async function readSingleBatch<T>(cursor: mongo.FindCursor<T>): Promise<{ data: T[]; hasMore: boolean }> {
  try {
    let data: T[];
    let hasMore = true;
    // Let MongoDB load the first batch of data
    const hasAny = await cursor.hasNext();
    // Now it's in memory, and we can read it
    data = cursor.readBufferedDocuments();
    if (!hasAny || cursor.id?.isZero()) {
      // A zero id means the cursor is exhaused.
      // No results (hasAny == false) means even this batch doesn't have data.
      // This should similar results as `await cursor.hasNext()`, but without
      // actually fetching the next batch.
      //
      // Note that it is safe (but slightly inefficient) to return `hasMore: true`
      // without there being more data, as long as the next batch
      // will return `hasMore: false`.
      hasMore = false;
    }
    return { data, hasMore };
  } finally {
    // Match the from the cursor iterator logic here:
    // https://github.com/mongodb/node-mongodb-native/blob/e02534e7d1c627bf50b85ca39f5995dbf165ad44/src/cursor/abstract_cursor.ts#L327-L331
    if (!cursor.closed) {
      await cursor.close();
    }
  }
}

export const BSON_DESERIALIZE_OPTIONS: bson.DeserializeOptions = {
  // use bigint instead of Long
  useBigInt64: true
};

export function mapOpEntry(row: BucketDataDocument): OplogEntry {
  if (row.op == 'PUT' || row.op == 'REMOVE') {
    return {
      op_id: timestampToOpId(row._id.o),
      op: row.op,
      object_type: row.table,
      object_id: row.row_id,
      checksum: Number(row.checksum),
      subkey: replicaIdToSubkey(row.source_table!, row.source_key!),
      data: row.data
    };
  } else {
    // MOVE, CLEAR

    return {
      op_id: timestampToOpId(row._id.o),
      op: row.op,
      checksum: Number(row.checksum)
    };
  }
}

/**
 * Returns true if two ReplicaId values are the same (serializes to the same BSON value).
 */
export function replicaIdEquals(a: ReplicaId, b: ReplicaId) {
  if (a === b) {
    return true;
  } else if (typeof a == 'string' && typeof b == 'string') {
    return a == b;
  } else if (isUUID(a) && isUUID(b)) {
    return a.equals(b);
  } else if (a == null && b == null) {
    return true;
  } else if (a != null || b != null) {
    return false;
  } else {
    // There are many possible primitive values, this covers them all
    return (bson.serialize({ id: a }) as Buffer).equals(bson.serialize({ id: b }));
  }
}

export function replicaIdToSubkey(table: bson.ObjectId, id: ReplicaId): string {
  if (isUUID(id)) {
    // Special case for UUID for backwards-compatiblity
    return `${table.toHexString()}/${id.toHexString()}`;
  } else {
    // Hashed UUID from the table and id
    const repr = bson.serialize({ table, id });
    return uuid.v5(repr, ID_NAMESPACE);
  }
}

/**
 * True if this is a bson.UUID.
 *
 * Works even with multiple copies of the bson package.
 */
export function isUUID(value: any): value is bson.UUID {
  if (value == null || typeof value != 'object') {
    return false;
  }
  const uuid = value as bson.UUID;
  return uuid._bsontype == 'Binary' && uuid.sub_type == bson.Binary.SUBTYPE_UUID;
}
