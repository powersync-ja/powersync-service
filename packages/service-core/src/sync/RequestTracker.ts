import { Metrics } from '../metrics/Metrics.js';

/**
 * Record sync stats per request stream.
 */
export class RequestTracker {
  operationsSynced = 0;
  dataSyncedBytes = 0;

  addOperationsSynced(operations: number) {
    this.operationsSynced += operations;

    Metrics.getInstance().operations_synced_total.add(operations);
  }

  addDataSynced(bytes: number) {
    this.dataSyncedBytes += bytes;

    Metrics.getInstance().data_synced_bytes.add(bytes);
  }
}
