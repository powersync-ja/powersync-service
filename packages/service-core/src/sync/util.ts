import * as timers from 'timers/promises';

import * as util from '../util/util-index.js';
import { Metrics } from '../metrics/Metrics.js';

export type TokenStreamOptions = {
  /**
   * Adds periodic keepalive events
   */
  keep_alive: boolean;
  /**
   * Warn before the token is going to expire
   */
  expire_warning_period: number;
};

const KEEPALIVE_INTERVAL = 20_000;

const DEFAULT_TOKEN_STREAM_OPTIONS: TokenStreamOptions = {
  keep_alive: true,
  expire_warning_period: 20_000
};

/**
 * An iterator that periodically yields token and optionally keepalive events, and returns once the
 * provided token expiry is reached.
 *
 * @param token exp is expiry as a unix timestamp (seconds)
 * @param signal abort the iterator with this
 * @param options configure keepalive and expire warnings
 */
export async function* tokenStream(
  token: { exp: number },
  signal: AbortSignal,
  options?: Partial<TokenStreamOptions>
): AsyncGenerator<util.StreamingSyncKeepalive> {
  const resolved_options: TokenStreamOptions = {
    ...DEFAULT_TOKEN_STREAM_OPTIONS,
    ...(options ?? {})
  };

  const { keep_alive, expire_warning_period } = resolved_options;

  // Real tokens always have an integer for exp.
  // In tests, we may use fractional seconds to reduce the delay.
  // Both cases are handled here.
  const expires_at = token.exp * 1000;
  const expire_warning_at = expires_at - expire_warning_period;

  let first_expire_test = true;

  while (!signal.aborted) {
    const token_expires_in = Math.max(0, Math.ceil(token.exp - Date.now() / 1000));
    if (first_expire_test && token_expires_in > 0) {
      first_expire_test = false;
    } else {
      yield { token_expires_in: token_expires_in };
      if (token_expires_in == 0) {
        return;
      }
    }

    const keep_alive_delay = KEEPALIVE_INTERVAL * (0.9 + Math.random() * 0.2);

    // Add margin due to setTimeout inaccuracies
    const expiry_delay = Math.max(0, expires_at - Date.now() + 3);
    const expiry_warning_delay = Math.max(0, expire_warning_at - Date.now() + 3);
    // Either the warning has past or it's before
    const relevant_expiry_delay = expiry_warning_delay != 0 ? expiry_warning_delay : expiry_delay;

    const delay = keep_alive ? Math.min(relevant_expiry_delay, keep_alive_delay) : relevant_expiry_delay;
    await timers.setTimeout(delay, null, { signal }).catch(() => {
      // Ignore AbortError
    });
  }
}

export async function* ndjson(iterator: AsyncIterable<string | null | Record<string, any>>): AsyncGenerator<string> {
  for await (let data of iterator) {
    if (data == null) {
      // Empty value to flush iterator memory
      continue;
    } else if (typeof data == 'string') {
      // Pre-serialized value
      yield data + '\n';
    } else {
      yield JSON.stringify(data) + '\n';
    }
  }
}

export async function* transformToBytesTracked(iterator: AsyncIterable<string>): AsyncGenerator<Buffer> {
  for await (let data of iterator) {
    const encoded = Buffer.from(data, 'utf8');
    Metrics.getInstance().data_synced_bytes.add(encoded.length);
    yield encoded;
  }
}
