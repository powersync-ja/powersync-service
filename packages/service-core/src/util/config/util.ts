import * as urijs from 'uri-js';

export function getSupabaseJwksUrl(connection: any) {
  if (connection == null) {
    return null;
  } else if (connection.type != 'postgresql') {
    return null;
  }

  let hostname: string | undefined = connection.hostname;
  if (hostname == null && typeof connection.uri == 'string') {
    hostname = urijs.parse(connection.uri).host;
  }
  if (hostname == null) {
    return null;
  }

  const match = /db.(\w+).supabase.co/.exec(hostname);
  if (match == null) {
    return null;
  }
  const projectId = match[1];

  return `https://${projectId}.supabase.co/auth/v1/.well-known/jwks.json`;
}
