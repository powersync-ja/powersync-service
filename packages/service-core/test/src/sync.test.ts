import { RequestTracker } from '@/sync/RequestTracker.js';
import { StreamingSyncLine } from '@/util/protocol-types.js';
import { lsnMakeComparable } from '@powersync/service-jpgwire';
import { JSONBig } from '@powersync/service-jsonbig';
import { RequestParameters } from '@powersync/service-sync-rules';
import * as timers from 'timers/promises';
import { describe, expect, test } from 'vitest';
import { ZERO_LSN } from '../../src/replication/WalStream.js';
import { streamResponse } from '../../src/sync/sync.js';
import { makeTestTable, MONGO_STORAGE_FACTORY, StorageFactory } from './util.js';

describe('sync - mongodb', function () {
  defineTests(MONGO_STORAGE_FACTORY);
});

const TEST_TABLE = makeTestTable('test', ['id']);

const BASIC_SYNC_RULES = `
bucket_definitions:
  mybucket:
    data:
      - SELECT * FROM test
    `;

function defineTests(factory: StorageFactory) {
  const tracker = new RequestTracker();

  test('sync global data', async () => {
    const f = await factory();

    const syncRules = await f.updateSyncRules({
      content: BASIC_SYNC_RULES
    });

    const storage = await f.getInstance(syncRules.parsed());
    await storage.setSnapshotDone(ZERO_LSN);
    await storage.autoActivate();

    const result = await storage.startBatch({}, async (batch) => {
      await batch.save({
        sourceTable: TEST_TABLE,
        tag: 'insert',
        after: {
          id: 't1',
          description: 'Test 1'
        }
      });

      await batch.save({
        sourceTable: TEST_TABLE,
        tag: 'insert',
        after: {
          id: 't2',
          description: 'Test 2'
        }
      });

      await batch.commit(lsnMakeComparable('0/1'));
    });

    const stream = streamResponse({
      storage: f,
      params: {
        buckets: [],
        include_checksum: true,
        raw_data: true
      },
      tracker,
      syncParams: new RequestParameters({ sub: '' }, {}),
      token: { exp: Date.now() / 1000 + 10 } as any
    });

    const lines = await consumeCheckpointLines(stream);
    expect(lines).toMatchSnapshot();
  });

  test('sync legacy non-raw data', async () => {
    const f = await factory();

    const syncRules = await f.updateSyncRules({
      content: BASIC_SYNC_RULES
    });

    const storage = await f.getInstance(syncRules.parsed());
    await storage.setSnapshotDone(ZERO_LSN);
    await storage.autoActivate();

    const result = await storage.startBatch({}, async (batch) => {
      await batch.save({
        sourceTable: TEST_TABLE,
        tag: 'insert',
        after: {
          id: 't1',
          description: 'Test\n"string"',
          large_num: 12345678901234567890n
        }
      });

      await batch.commit(lsnMakeComparable('0/1'));
    });

    const stream = streamResponse({
      storage: f,
      params: {
        buckets: [],
        include_checksum: true,
        raw_data: false
      },
      tracker,
      syncParams: new RequestParameters({ sub: '' }, {}),
      token: { exp: Date.now() / 1000 + 10 } as any
    });

    const lines = await consumeCheckpointLines(stream);
    expect(lines).toMatchSnapshot();
    // Specifically check the number - this is the important part of the test
    expect(lines[1].data.data[0].data.large_num).toEqual(12345678901234567890n);
  });

  test('expired token', async () => {
    const f = await factory();

    const syncRules = await f.updateSyncRules({
      content: BASIC_SYNC_RULES
    });

    const storage = await f.getInstance(syncRules.parsed());
    await storage.setSnapshotDone(ZERO_LSN);
    await storage.autoActivate();

    const stream = streamResponse({
      storage: f,
      params: {
        buckets: [],
        include_checksum: true,
        raw_data: true
      },
      tracker,
      syncParams: new RequestParameters({ sub: '' }, {}),
      token: { exp: 0 } as any
    });

    const lines = await consumeCheckpointLines(stream);
    expect(lines).toMatchSnapshot();
  });

  test('sync updates to global data', async () => {
    const f = await factory();

    const syncRules = await f.updateSyncRules({
      content: BASIC_SYNC_RULES
    });

    const storage = await f.getInstance(syncRules.parsed());
    await storage.setSnapshotDone(ZERO_LSN);
    await storage.autoActivate();

    const stream = streamResponse({
      storage: f,
      params: {
        buckets: [],
        include_checksum: true,
        raw_data: true
      },
      tracker,
      syncParams: new RequestParameters({ sub: '' }, {}),
      token: { exp: Date.now() / 1000 + 10 } as any
    });
    const iter = stream[Symbol.asyncIterator]();

    expect(await getCheckpointLines(iter)).toMatchSnapshot();

    await storage.startBatch({}, async (batch) => {
      await batch.save({
        sourceTable: TEST_TABLE,
        tag: 'insert',
        after: {
          id: 't1',
          description: 'Test 1'
        }
      });

      await batch.commit(lsnMakeComparable('0/1'));
    });

    expect(await getCheckpointLines(iter)).toMatchSnapshot();

    await storage.startBatch({}, async (batch) => {
      await batch.save({
        sourceTable: TEST_TABLE,
        tag: 'insert',
        after: {
          id: 't2',
          description: 'Test 2'
        }
      });

      await batch.commit(lsnMakeComparable('0/2'));
    });

    expect(await getCheckpointLines(iter)).toMatchSnapshot();

    iter.return?.();
  });

  test('expiring token', async () => {
    const f = await factory();

    const syncRules = await f.updateSyncRules({
      content: BASIC_SYNC_RULES
    });

    const storage = await f.getInstance(syncRules.parsed());
    await storage.setSnapshotDone(ZERO_LSN);
    await storage.autoActivate();

    const exp = Date.now() / 1000 + 0.1;

    const stream = streamResponse({
      storage: f,
      params: {
        buckets: [],
        include_checksum: true,
        raw_data: true
      },
      tracker,
      syncParams: new RequestParameters({ sub: '' }, {}),
      token: { exp: exp } as any
    });
    const iter = stream[Symbol.asyncIterator]();

    const checkpoint = await getCheckpointLines(iter);
    expect(checkpoint).toMatchSnapshot();

    const expLines = await getCheckpointLines(iter);
    expect(expLines).toMatchSnapshot();
  });

  test('compacting data - invalidate checkpoint', async () => {
    // This tests a case of a compact operation invalidating a checkpoint in the
    // middle of syncing data.
    // This is expected to be rare in practice, but it is important to handle
    // this case correctly to maintain consistency on the client.

    const f = await factory();

    const syncRules = await f.updateSyncRules({
      content: BASIC_SYNC_RULES
    });

    const storage = await f.getInstance(syncRules.parsed());
    await storage.setSnapshotDone(ZERO_LSN);
    await storage.autoActivate();

    await storage.startBatch({}, async (batch) => {
      await batch.save({
        sourceTable: TEST_TABLE,
        tag: 'insert',
        after: {
          id: 't1',
          description: 'Test 1'
        }
      });

      await batch.save({
        sourceTable: TEST_TABLE,
        tag: 'insert',
        after: {
          id: 't2',
          description: 'Test 2'
        }
      });

      await batch.commit(lsnMakeComparable('0/1'));
    });

    const stream = streamResponse({
      storage: f,
      params: {
        buckets: [],
        include_checksum: true,
        raw_data: true
      },
      tracker,
      syncParams: new RequestParameters({ sub: '' }, {}),
      token: { exp: Date.now() / 1000 + 10 } as any
    });

    const iter = stream[Symbol.asyncIterator]();

    // Only consume the first "checkpoint" message, and pause before receiving data.
    const lines = await consumeIterator(iter, { consume: false, isDone: (line) => (line as any)?.checkpoint != null });
    expect(lines).toMatchSnapshot();
    expect(lines[0]).toEqual({
      checkpoint: expect.objectContaining({
        last_op_id: '2'
      })
    });

    // Now we save additional data AND compact before continuing.
    // This invalidates the checkpoint we've received above.

    await storage.startBatch({}, async (batch) => {
      await batch.save({
        sourceTable: TEST_TABLE,
        tag: 'update',
        after: {
          id: 't1',
          description: 'Test 1b'
        }
      });

      await batch.save({
        sourceTable: TEST_TABLE,
        tag: 'update',
        after: {
          id: 't2',
          description: 'Test 2b'
        }
      });

      await batch.commit(lsnMakeComparable('0/2'));
    });

    await storage.compact();

    const lines2 = await getCheckpointLines(iter, { consume: true });

    // Snapshot test checks for changes in general.
    // The tests after that documents the specific things we're looking for
    // in this test.
    expect(lines2).toMatchSnapshot();

    expect(lines2[0]).toEqual({
      data: expect.objectContaining({
        has_more: false,
        data: [
          // The first two ops have been replaced by a single CLEAR op
          expect.objectContaining({
            op: 'CLEAR'
          })
        ]
      })
    });

    // Note: No checkpoint_complete here, since the checkpoint has been
    // invalidated by the CLEAR op.

    expect(lines2[1]).toEqual({
      checkpoint_diff: expect.objectContaining({
        last_op_id: '4'
      })
    });

    expect(lines2[2]).toEqual({
      data: expect.objectContaining({
        has_more: false,
        data: [
          expect.objectContaining({
            op: 'PUT'
          }),
          expect.objectContaining({
            op: 'PUT'
          })
        ]
      })
    });

    // Now we get a checkpoint_complete
    expect(lines2[3]).toEqual({
      checkpoint_complete: expect.objectContaining({
        last_op_id: '4'
      })
    });
  });
}

/**
 * Get lines on an iterator until isDone(line) == true.
 *
 * Does not stop the iterator unless options.consume is true.
 */
async function consumeIterator<T>(
  iter: AsyncIterator<T>,
  options: { isDone: (line: T) => boolean; consume?: boolean }
) {
  let lines: T[] = [];
  try {
    const controller = new AbortController();
    const timeout = timers.setTimeout(1500, { value: null, done: 'timeout' }, { signal: controller.signal });
    while (true) {
      let { value, done } = await Promise.race([timeout, iter.next()]);
      if (done == 'timeout') {
        throw new Error('Timeout');
      }
      if (typeof value == 'string') {
        value = JSONBig.parse(value);
      }
      if (value) {
        lines.push(value);
      }
      if (done || options.isDone(value)) {
        break;
      }
    }
    controller.abort();

    if (options?.consume) {
      iter.return?.();
    }
    return lines;
  } catch (e) {
    if (options?.consume) {
      iter.throw?.(e);
    }
    throw e;
  }
}

/**
 * Get lines on an iterator until the next checkpoint_complete.
 *
 * Does not stop the iterator unless options.consume is true.
 */
async function getCheckpointLines(
  iter: AsyncIterator<StreamingSyncLine | string | null>,
  options?: { consume?: boolean }
) {
  return consumeIterator(iter, {
    consume: options?.consume,
    isDone: (line) => (line as any)?.checkpoint_complete
  });
}

/**
 * Get lines on an iterator until the next checkpoint_complete.
 *
 * Stops the iterator afterwards.
 */
async function consumeCheckpointLines(iterable: AsyncIterable<StreamingSyncLine | string | null>): Promise<any[]> {
  return getCheckpointLines(iterable[Symbol.asyncIterator](), { consume: true });
}
