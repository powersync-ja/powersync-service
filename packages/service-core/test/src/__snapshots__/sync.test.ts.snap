// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`sync - mongodb > expired token 1`] = `
[
  {
    "token_expires_in": 0,
  },
]
`;

exports[`sync - mongodb > expiring token 1`] = `
[
  {
    "checkpoint": {
      "buckets": [
        {
          "bucket": "mybucket[]",
          "checksum": 0,
          "count": 0,
        },
      ],
      "last_op_id": "0",
      "write_checkpoint": undefined,
    },
  },
  {
    "checkpoint_complete": {
      "last_op_id": "0",
    },
  },
]
`;

exports[`sync - mongodb > expiring token 2`] = `
[
  {
    "token_expires_in": 0,
  },
]
`;

exports[`sync - mongodb > sync global data 1`] = `
[
  {
    "checkpoint": {
      "buckets": [
        {
          "bucket": "mybucket[]",
          "checksum": -93886621,
          "count": 2,
        },
      ],
      "last_op_id": "2",
      "write_checkpoint": undefined,
    },
  },
  {
    "data": {
      "after": "0",
      "bucket": "mybucket[]",
      "data": [
        {
          "checksum": 920318466n,
          "data": "{\\"id\\":\\"t1\\",\\"description\\":\\"Test 1\\"}",
          "object_id": "t1",
          "object_type": "test",
          "op": "PUT",
          "op_id": "1",
          "subkey": "6544e3899293153fa7b38331/117ab485-4b42-58a2-ab32-0053a22c3423",
        },
        {
          "checksum": 3280762209n,
          "data": "{\\"id\\":\\"t2\\",\\"description\\":\\"Test 2\\"}",
          "object_id": "t2",
          "object_type": "test",
          "op": "PUT",
          "op_id": "2",
          "subkey": "6544e3899293153fa7b38331/ec27c691-b47a-5d92-927a-9944feb89eee",
        },
      ],
      "has_more": false,
      "next_after": "2",
    },
  },
  {
    "checkpoint_complete": {
      "last_op_id": "2",
    },
  },
]
`;

exports[`sync - mongodb > sync legacy non-raw data 1`] = `
[
  {
    "checkpoint": {
      "buckets": [
        {
          "bucket": "mybucket[]",
          "checksum": -852817836,
          "count": 1,
        },
      ],
      "last_op_id": "1",
      "write_checkpoint": undefined,
    },
  },
  {
    "data": {
      "after": "0",
      "bucket": "mybucket[]",
      "data": [
        {
          "checksum": 3442149460n,
          "data": {
            "description": "Test
\\"string\\"",
            "id": "t1",
            "large_num": 12345678901234567890n,
          },
          "object_id": "t1",
          "object_type": "test",
          "op": "PUT",
          "op_id": "1",
          "subkey": "6544e3899293153fa7b38331/117ab485-4b42-58a2-ab32-0053a22c3423",
        },
      ],
      "has_more": false,
      "next_after": "1",
    },
  },
  {
    "checkpoint_complete": {
      "last_op_id": "1",
    },
  },
]
`;

exports[`sync - mongodb > sync updates to global data 1`] = `
[
  {
    "checkpoint": {
      "buckets": [
        {
          "bucket": "mybucket[]",
          "checksum": 0,
          "count": 0,
        },
      ],
      "last_op_id": "0",
      "write_checkpoint": undefined,
    },
  },
  {
    "checkpoint_complete": {
      "last_op_id": "0",
    },
  },
]
`;

exports[`sync - mongodb > sync updates to global data 2`] = `
[
  {
    "checkpoint_diff": {
      "last_op_id": "1",
      "removed_buckets": [],
      "updated_buckets": [
        {
          "bucket": "mybucket[]",
          "checksum": 920318466,
          "count": 1,
        },
      ],
      "write_checkpoint": undefined,
    },
  },
  {
    "data": {
      "after": "0",
      "bucket": "mybucket[]",
      "data": [
        {
          "checksum": 920318466n,
          "data": "{\\"id\\":\\"t1\\",\\"description\\":\\"Test 1\\"}",
          "object_id": "t1",
          "object_type": "test",
          "op": "PUT",
          "op_id": "1",
          "subkey": "6544e3899293153fa7b38331/117ab485-4b42-58a2-ab32-0053a22c3423",
        },
      ],
      "has_more": false,
      "next_after": "1",
    },
  },
  {
    "checkpoint_complete": {
      "last_op_id": "1",
    },
  },
]
`;

exports[`sync - mongodb > sync updates to global data 3`] = `
[
  {
    "checkpoint_diff": {
      "last_op_id": "2",
      "removed_buckets": [],
      "updated_buckets": [
        {
          "bucket": "mybucket[]",
          "checksum": -93886621,
          "count": 2,
        },
      ],
      "write_checkpoint": undefined,
    },
  },
  {
    "data": {
      "after": "1",
      "bucket": "mybucket[]",
      "data": [
        {
          "checksum": 3280762209n,
          "data": "{\\"id\\":\\"t2\\",\\"description\\":\\"Test 2\\"}",
          "object_id": "t2",
          "object_type": "test",
          "op": "PUT",
          "op_id": "2",
          "subkey": "6544e3899293153fa7b38331/ec27c691-b47a-5d92-927a-9944feb89eee",
        },
      ],
      "has_more": false,
      "next_after": "2",
    },
  },
  {
    "checkpoint_complete": {
      "last_op_id": "2",
    },
  },
]
`;
